### ISSUE TITLE
Edge weights of the k-nearest-neighbor graph

### ISSUE THREAD

**Issue by katjaschwarz** [lmcinnes/umap#164]
Hello Leland,

First: UMAP is amazing, thank you for making the code publicly available and putting so much effort in what is going on there!

If I understood the "read the docs" correctly, UMAP effectively computes a weighted k-nearest neighbor graph from the data. The edge weights to the k-nearest neighbors from a data point x_i consider the local metric around the point, which is done using the distance relative with respect to its closest neighbor. I was wondering if there is an explicit formula for computing the edge weights? 

At https://www.math.upenn.edu/~jhansen/2018/05/04/UMAP/ they state, that the weights can be computed as 

w_i(x_i, x_j) = exp(-(d(x_i, x_j) - r_i) / sigma_i), 

where r_i is the distance to the closest neighbor and sigma_i is the diameter of the neighborhood. Is this formula correct? If yes, how one compute the edge weights for the low-dimensional representation, since we do not use neighbors, but rather the hyperparameter min_dist?
Also I did not quite understand how the family of curves 1/(1+ax^(2b)) relates to all this and I would really appreciate your help.

Cheers,

Katja

**Comment by katjaschwarz** [comment]
Yes, this fully answered my question.
Thanks a lot for the fast and helpful reply!!

**Comment by qbit12** [comment]
is there a loss function thats minimized using gradient descent over here when constructing weighted knn graph
